src/coreclr/vm/jithelpers.cpp

```diff
@@ -2431,7 +2432,19 @@ OBJECTHANDLE ConstructStringLiteral(CORINFO_MODULE_HANDLE scopeHnd, mdToken meta
     _ASSERTE(TypeFromToken(metaTok) == mdtString);
 
     Module* module = GetModule(scopeHnd);
-    return module->ResolveStringRef(metaTok);
+
+    // ResolveStringRef returns a small pinned object that points to StringObject
+    OBJECTHANDLE strObjHandle = module->ResolveStringRef(metaTok);
+
+    // Let's see if that StringObject is pinned too (stored in Frozen Object Heap)
+    FrozenObjectHeap* foh = SystemDomain::GetSegmentWithFrozenObjects();
```

Comment:
This logic should be in `StringLiteralMap`. The `ppPinnedString` argument should be passed through the several layers.